### PR TITLE
[FLINK-26808] Limit FileUploadHandler to multipart routes

### PR DESCRIPTION
**Pull request by uce**
## What is the purpose of the change

The purpose of this change is to limit multipart requests to POST endpoints and file upload endpoint. Prior to this change, any request could be a multipart request (and upload a file).

With this change, the behavior of `FileUploadHandler` will be as follows:
1. Only accept a multipart request if the endpoint handler accepts POST requests.
2. Only accept a file upload if the endpoint handler accept file uploads.

Note that there multipart requests can mix file uploads with regular requests.

## Brief change log

  - Add `MultipartRoutes` in the `*.router` package for re-use of existing path patterns utilities.
  - Factor out `RestServerEndpoint#getHandlerUrls(RestHandlerSpecification)` in order to ensure that we have a consistent way of getting the registered endpoint URLs.
  - Add `MultipartRoutes` to `FileUploadHandler` and populate from handler list in `RestServerEndpoint`.
  - Validate that all multipart requests go to a POST endpoint.
  - Validate that file uploads go to a endpoint that accepts file uploads.

## Verifying this change

This change is already covered by existing tests, such as `RestServerEndpointITCase` and `FileUploadHandlerITCase`. I made minor additions to these tests.

I manually verified that I can submit jobs via `bin/flink run` and `curl` both when web sumission is enabled and disabled.

The following gets logged in DEBUG on startup:

When web submission enabled:

```
DEBUG org.apache.flink.runtime.dispatcher.DispatcherRestEndpoint   [] - Using MultipartRoutes{postRoutes=[v1/jars/upload, jars/upload, v1/jars/:jarid/plan, jars/:jarid/plan, v1/jars/:jarid/run, jars/:jarid/run, v1/jobs, jobs, v1/jobs/:jobid/checkpoints, jobs/:jobid/checkpoints, v1/jobs/:jobid/coordinators/:operatorid, jobs/:jobid/coordinators/:operatorid, v1/jobs/:jobid/savepoints, jobs/:jobid/savepoints, v1/jobs/:jobid/stop, jobs/:jobid/stop, v1/savepoint-disposal, savepoint-disposal], fileUploadRoutes=[v1/jars/upload, jars/upload, v1/jobs, jobs]} for FileUploadHandler
```

When web submission disabled:

```
DEBUG org.apache.flink.runtime.dispatcher.DispatcherRestEndpoint   [] - Using MultipartRoutes{postRoutes=[v1/jobs, jobs, v1/jobs/:jobid/checkpoints, jobs/:jobid/checkpoints, v1/jobs/:jobid/coordinators/:operatorid, jobs/:jobid/coordinators/:operatorid, v1/jobs/:jobid/savepoints, jobs/:jobid/savepoints, v1/jobs/:jobid/stop, jobs/:jobid/stop, v1/savepoint-disposal, savepoint-disposal], fileUploadRoutes=[v1/jobs, jobs]} for FileUploadHandler
```

When trying to upload a JAR with web submission disabled:

```
$ curl -X POST -H "Expect:" -F "jarfile=@examples/streaming/StateMachineExample.jar" http://localhost:8081/jars/upload
{"errors":["POST request not allowed"]}
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: yes (REST API for job and JAR submission)
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Review by zentol — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutes.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package org.apache.flink.runtime.rest.handler.router;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutes.java:1]: missing license

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutesTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package org.apache.flink.runtime.rest.handler.router;
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutesTest.java:1]: missing license

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutes.java (hunk 1 above)
Comment (uce) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutes.java:19]: Fixed in 019f2ddad50.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutesTest.java (hunk 2 above)
Comment (uce) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/router/MultipartRoutesTest.java:19]: Fixed in 019f2ddad50.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java [hunk 3]
```diff
@@ -151,6 +165,17 @@ protected void channelRead0(final ChannelHandlerContext ctx, final HttpObject ms
                         && hasNext(currentHttpPostRequestDecoder)) {
                     final InterfaceHttpData data = currentHttpPostRequestDecoder.next();
                     if (data.getHttpDataType() == InterfaceHttpData.HttpDataType.FileUpload) {
+                        HttpRequest httpRequest = currentHttpRequest;
+                        if (!multipartRoutes.isFileUploadRoute(httpRequest.uri())) {
```
[the remark this answers is not in the document]
Comment (uce) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java:169]: We create a directory per upload and `handleError` deletes the current upload directory.

https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java#L225-L251

[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java:169]: sweet 👍

**Comment by uce** [comment]
I was too trigger happy and missed to fixup the `review` commit before merging. 😢

**Pull request merged**